### PR TITLE
Bug 2055814: Support custom extensions in console-extensions.json files

### PR DESCRIPTION
**Pull request by vojtechszocs**
Changes to dynamic plugin extensions schema code:

- add type `CustomExtension` that supports arbitrary `type` and `properties`
```ts
export type CustomExtension = ExtensionDeclaration<string, { [key: string]: any }>;
```
- update type `ConsoleExtensionsJSON` representing the schema of `console-extensions.json`
```diff
-export type ConsoleExtensionsJSON = SupportedExtension[];
+export type ConsoleExtensionsJSON = (SupportedExtension | CustomExtension)[];
```
- update code to work with `ConsoleExtensionsJSON` instead of `SupportedExtension[]`
- update `ExtensionDeclarationParser` to recognize `CustomExtension`

---

With these changes, the generated `console-extensions.json` schema is modified like so:

```diff
 {
   "$schema": "http://json-schema.org/draft-07/schema#",
   "type": "array",
   "items": {
-    "$ref": "#/definitions/SupportedExtension"
+    "anyOf": [
+      {
+        "$ref": "#/definitions/SupportedExtension"
+      },
+      {
+        "$ref": "#/definitions/CustomExtension"
+      }
+    ]
   },
   "description": "Schema of Console plugin's `console-extensions.json` file.",
   "definitions": {
     ...
     ... schema definitions for all SupportedExtension union members
     ...
+    "CustomExtension": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "type": {
+          "type": "string"
+        },
+        "properties": {
+          "type": "object"
+        },
+        "flags": {
+          "$ref": "#/definitions/ExtensionFlags"
+        }
+      },
+      "required": [
+        "properties",
+        "type"
+      ],
+      "description": "Custom Console extension with arbitrary `type` and `properties`.\n\nThis type represents all non-standard Console extension declarations."
+    }
   }
 }
```

---

Changes to dynamic demo plugin:
- add `.gitignore` to un-ignore the plugin's `.vscode` directory
- add `.vscode/settings.json` that configures `console-extensions.json` schema
- add symlink `.prettierrc` => `../frontend/.prettierrc`
- update outdated `yarn.lock` (`rm -rf node_modules ; yarn install`)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,6 +1,6 @@
 .env
-.idea
-.vscode
+/.idea
+/.vscode
```
Comment (andrewballantyne) [inline .gitignore:3]: What's the purpose behind this change? Either we want configs or we don't from IDEs... right?

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/custom.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { ExtensionDeclaration } from '../types';
+
+/**
+ * Custom Console extension with arbitrary `type` and `properties`.
+ *
+ * This type represents all non-standard Console extension declarations.
```
Comment (andrewballantyne) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/custom.ts:6]: Your PR doesn't describe what this is solving or link out to a ticket. Can you help me understand why we need this?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/custom.ts (hunk 2 above)
Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/custom.ts:6]: I'd mention this is so plugins can define their own extension points for other plugins to use.

File: .gitignore (hunk 1 above)
Comment (vojtechszocs) [reply .gitignore]: I've reverted those changes.

Generally the intention is to git-ignore all IDE specific config files.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/custom.ts (hunk 2 above)
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/custom.ts]: Done, the `CustomExtension` type doc should be better now :smiley:

**Comment by christianvogt** [comment]
lgtm

have @bipuladh give this a try

**Review by spadgett — APPROVED** [review]
/lgtm
/hold

We'll wait for @bipuladh's feedback

**Comment by andrewballantyne** [comment]
Seems you have a lint error @vojtechszocs 

```
/go/src/github.com/openshift/console/frontend/packages/console-dynamic-plugin-sdk/src/schema/console-extensions.ts
  66:1  error  `../extensions/custom` import should occur before import of `../extensions/dashboards`  import/order

. . .

✖ 155 problems (1 error, 154 warnings)
 1 error and 0 warnings potentially fixable with the `--fix` option.
 ```

**Comment by vojtechszocs** [comment]
> Seems you have a lint error @vojtechszocs

It should be fixed now :smiley:

**Review by spadgett — APPROVED** [review]
/lgtm

**Comment by bipuladh** [comment]
/lgtm
Tested with odf-console

Screenshot: 
Screenshot from 2022-02-17 22-28-46

**Comment by bipuladh** [comment]
/retitle Bug 2055814: Support custom extensions in console-extensions.json files
